packages/netutil/buffconn/buffconn.go

```diff
@@ -0,0 +1,172 @@
+	c.closeOnce.Do(func() {
+		err = c.conn.Close()
+		// close in separate go routine to avoid deadlocks
+		go c.Events.Close.Trigger()
+	})
+	return err
+}
+
+// LocalAddr returns the local network address.
+func (c *BufferedConnection) LocalAddr() net.Addr {
+	return c.conn.LocalAddr()
+}
+
+// RemoteAddr returns the remote network address.
+func (c *BufferedConnection) RemoteAddr() net.Addr {
+	return c.conn.RemoteAddr()
+}
+
+// Read starts reading on the connection, it only returns when an error occurred or when Close has been called.
+// If a complete message has been received and ReceiveMessage event is triggered with its complete payload.
+// If read leads to an error, the loop will be stopped and that error returned.
+func (c *BufferedConnection) Read() error {
+	buffer := make([]byte, MaxMessageSize)
+
+	for {
+		n, err := c.readMessage(buffer)
+		if err != nil {
+			return err
+		}
+		if n > 0 {
+			c.Events.ReceiveMessage.Trigger(buffer[:n])
+		}
+	}
+}
+
+// Write sends a stream of bytes as messages.
+// Each array of bytes you pass in will be pre-pended with it's size. If the
+// connection isn't open you will receive an error. If not all bytes can be
+// written, Write will keep trying until the full message is delivered, or the
+// connection is broken.
+func (c *BufferedConnection) Write(msg []byte) (int, error) {
+	if l := len(msg); l > MaxMessageSize {
+		panic(fmt.Sprintf("invalid message length: %d", l))
+	}
+
+	buffer := append(newHeader(len(msg)), msg...)
+
+	if err := c.conn.SetWriteDeadline(time.Now().Add(IOTimeout)); err != nil {
+		return 0, fmt.Errorf("error while setting timeout: %w", err)
+	}
+
+	toWrite := len(buffer)
+	for bytesWritten := 0; bytesWritten < toWrite; {
+		n, err := c.conn.Write(buffer[bytesWritten:])
+		bytesWritten += n
+		if err != nil {
+			return bytesWritten, err
+		}
+	}
+	return toWrite - headerSize, nil
+}
+
+func (c *BufferedConnection) read(buffer []byte) (int, error) {
+	toRead := len(buffer)
+	for bytesRead := 0; bytesRead < toRead; {
+		n, err := c.conn.Read(buffer[bytesRead:])
+		bytesRead += n
+		if err != nil {
+			return bytesRead, err
+		}
+	}
+	return toRead, nil
+}
+
+func (c *BufferedConnection) readMessage(buffer []byte) (int, error) {
+	if err := c.conn.SetReadDeadline(time.Time{}); err != nil {
+		return 0, fmt.Errorf("error while unsetting timeout: %w", err)
+	}
+	_, err := c.read(c.incomingHeaderBuffer)
+	if err != nil {
+		return 0, err
+	}
+
+	msgLength, err := parseHeader(c.incomingHeaderBuffer)
+	if err != nil {
+		return 0, err
+	}
+	if msgLength > len(buffer) {
+		return 0, ErrInsufficientBuffer
+	}
+
+	if err := c.conn.SetReadDeadline(time.Now().Add(IOTimeout)); err != nil {
+		return 0, fmt.Errorf("error while setting timeout: %w", err)
+	}
+	return c.read(buffer[:msgLength])
+}
+
+func newHeader(msgLength int) []byte {
+	// the header only consists of the message length
+	header := make([]byte, headerSize)
+	binary.BigEndian.PutUint32(header, uint32(msgLength))
```

Comment:
Since a message can be max 4kb, we can reduce this to an uint16?

Reply:
My reasoning for this was that maybe 65535 bytes could be a bit small in the future. If you reserve 32bits for the size in the header it might be easier to increase the size in the future without the need to change the header.
I don't know. What do you think does it make change to reserve this or not?

Reply:
Aight, fine by me then, after all it is just 2 bytes more. 
I just didn't really know we'd have single messages over 65kb in the future.

Reply:
was thinking of maybe atomic transactions :stuck_out_tongue_winking_eye: